micro-rdk/src/esp32/camera.rs

```diff
@@ -1,117 +1,254 @@
-                pin_d7: 35,
-                pin_d6: 34,
-                pin_d5: 39,
-                pin_d4: 36,
-                pin_d3: 19,
-                pin_d2: 18,
-                pin_d1: 5,
-                pin_d0: 4,
-                pin_vsync: 25,
-                pin_href: 23,
-                pin_pclk: 22,
-                xclk_freq_hz: 20000000,
-                ledc_timer: 1,
-                ledc_channel: 1,
-                pixel_format: 4,
-                frame_size: 4,
-                jpeg_quality: 32,
+                pin_pwdn,
+                pin_reset,
+                pin_xclk,
+                __bindgen_anon_1: camera_config_t__bindgen_ty_1 { pin_sccb_sda },
+                __bindgen_anon_2: camera_config_t__bindgen_ty_2 { pin_sccb_scl },
+                pin_d7,
+                pin_d6,
+                pin_d5,
+                pin_d4,
+                pin_d3,
+                pin_d2,
+                pin_d1,
+                pin_d0,
+                pin_vsync,
+                pin_href,
+                pin_pclk,
+                xclk_freq_hz,
+                ledc_channel,
+                ledc_timer,
+                pixel_format: PixelFormat::JPEG as u32,
+                frame_size,
+                jpeg_quality,
+                // Number of frame buffers to be allocated.
+                // If more than one, then each frame will be acquired (double speed)
+                // when pixel_format == jpeg, fb_count > 1 goes to continuous mode, may need to adjust
+                // xclk_freq_hz down to 10_000_000.
                 fb_count: 1,
                 grab_mode: 0,
                 fb_location: 0,
-                sccb_i2c_port: 0,
+                sccb_i2c_port,
             },
-            last_grab: t.now(),
-        }
+        };
+
+        esp!(unsafe { esp_camera_init(&cam.config) })
+            .map_err(|e| CameraError::InitError(Box::new(e)))?;
+        Ok(Arc::new(Mutex::new(cam)))
     }
-    pub fn setup(&self) -> Result<(), CameraError> {
-        let ret = (unsafe { crate::esp32::esp_idf_svc::sys::esp_camera_init(&self.config) })
-            as crate::esp32::esp_idf_svc::sys::esp_err_t;
-        let ret = crate::esp32::esp_idf_svc::sys::EspError::convert(ret);
-        ret.map_err(|e| CameraError::CameraInitError(e.into()))
+
+    fn get_frame(&mut self) -> Result<Esp32CameraFrameBuffer, CameraError> {
+        Esp32CameraFrameBuffer::get().ok_or(CameraError::FailedToGetImage)
     }
-    pub fn get_cam_frame(&self) -> Option<*mut crate::esp32::esp_idf_svc::sys::camera_fb_t> {
-        let ptr = (unsafe { crate::esp32::esp_idf_svc::sys::esp_camera_fb_get() })
-            as *mut crate::esp32::esp_idf_svc::sys::camera_fb_t;
-        if ptr.is_null() {
-            None
-        } else {
-            Some(ptr)
+}
+
+impl Camera for Esp32Camera {
+    fn get_image(&mut self, mut buffer: BytesMut) -> Result<BytesMut, CameraError> {
```

Comment:
How does the caller know how big a buffer to provide to avoid the error or be able to succeed on the next attempt?

Reply:
This is a good point. The actual underlying buffer has its size hard-coded. The user is indirectly "requesting" a specific buffer size depending on the resolution they choose. So, a camera that has been set up with `FrameSize:: W240XH240` will work with the current hardcoding, but one initialized with `FrameSize::VGA` would not fit within the buffer, and we would know this before the first `get_image` request.

I think the best option for now is to error on the camera creation if the `FrameSize` exceeds the currently configured buffer. This will let them choose another `FrameSize` in app that will work.

Reply:
This would probably make the capacity check unnecessary as well

Reply:
Let's talk this one over in person when you have a minute. I may be misunderstanding something about how this API is supposed to work, and I don't want to waste your time on it.

Reply:
discussed offline with @acmorrow. The current grpc/webrtc buffer is statically sized, limiting the usable `frame_size`s and requiring these checks.

Created a follow up ticket: [RSDK-8830](https://viam.atlassian.net/browse/RSDK-8330?atlOrigin=eyJpIjoiOTU1NzM5N2ZhZDdjNDg3NTlhM2JmZTVlNWZiMTRlOGQiLCJwIjoiaiJ9)